docs/modules/candid-guide/index.adoc

```diff
@@ -0,0 +1,1098 @@
+
+Textual syntax::
+
+`blob <text>`, where `<text>` is a text literal (see <<type-text,above>>), with all characters representing their utf8 encoding, and arbitray byte sequences (`"\CA\FF\FE"`).
+
+Subtypes::
+
+`vec nat8`, and all subtypes of that.
+
+Supertypes::
+
+`vec nat8`, and all supertypes of that.
+
+Corresponding Motoko type::
+
+`Blob`
+
+Corresponding Rust type::
+
+`Vec<u8>`
+
+Corresponding JavaScript values::
+
+TODO:
+
+[#type-nat]
+=== Type nat
+
+The `nat` type contains all natural (non-negative) numbers. It is unbounded, and can represent arbitrary large numbers.  The on-wire encoding is LEB128, so small numbers are still efficiently represented.
+
+Type syntax::
+
+`nat`
+
+Textual syntax::
++
+....
+1234
+1_000_000
+0xDEAD_BEEF
+....
+
+
+Supertypes::
+
+`int`
+
+Corresponding Motoko type::
+
+`Nat`
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+
+[#type-int]
+=== Type int
+
+The `int` type contains all integral, i.e. whole, numbers. It is unbounded, and can represent arbitrary small or large numbers. The on-wire encoding is SLEB128, so small numbers are still efficiently represented.
+
+Type syntax::
+
+`int`
+
+Textual syntax::
++
+....
+1234
+-1234
++1234
+1_000_000
+-1_000_000
++1_000_000
+0xDEAD_BEEF
+-0xDEAD_BEEF
++0xDEAD_BEEF
+....
+
+Subtypes::
+
+`nat`
+
+Corresponding Motoko type::
+
+`Int`
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+
+[#type-natN]
+[#type-intN]
+=== Type natN and intN
+
+The types `nat8`, `nat16`, `nat32`, `nat64`, `int8`, `int16`, `int32` and `nat64` represent numbers with a representation of that many bits, and can be used in more “low-level” interface.
+
+The range of `natN` is \(\{0 \ldots 2^{N}-1\}\), and the range of `intN` is \(\{-2^{N-1} \ldots 2^{N-1}-1\}\).
+
+The on-wire representation is exactly that many bits long. So for small values, `nat` is more space-efficent than `nat64`.
+
+Type syntax::
+
+`nat8`, `nat16`, `nat32`, `nat64`, `int8`, `int16`, `int32` or `nat64`
+
+Textual syntax::
+
+Same as `nat` (for `natN`) resp. `int` (for `intN`).
+
+Corresponding Motoko type::
+
+`natN` translate by default to `NatN`, but can also correspond got `WordN`
++
+`intN` translate to `IntN`.
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+[#type-floatN]
+=== Type float16 and float32
+
+The types `float16`, `float32` represent IEEE 754 floating point numbers, in single precision (32 bit) and double precision (64 bit).
+
+Type syntax::
+
+`float16`, `float32`
+
+Textual syntax::
+
+The same syntax as `int`, plus floating point literals as follows
++
+....
+1245.678
++1245.678
+-1_000_000.000_001
+34e10
+34E+10
+34e-10
+0xDEAD.BEEF
+0xDEAD.BEEFP-10
+0xDEAD.BEEFp+10
+....
+
+Corresponding Motoko type::
+
+`float64` corresponds to `Float`.
++
+`float32` does _not_ have a representation in Motoko. Candid interfaces using `float32` cannot be served from or used from Motoko programs.
```

Comment:
Maybe we should fix that...

Reply:
Isn’t it nice how writing documentation really makes you feel the pain of certain design choices? Almost as good as actually using the stuff

Reply:
Yeah, highly recommend both writing doc and dog-fooding...